File: Temp_AFPGOCAResources/src/java/org/apache/fop/pdf/PDFObject.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* $Id$ */

package org.apache.fop.pdf;

// Java
import java.io.IOException;
import java.io.OutputStream;
import java.io.Writer;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

/**
 * generic PDF object.
 *
 * A PDF Document is essentially a collection of these objects. A PDF
 * Object has a number and a generation (although the generation will always
 * be 0 in new documents).
 */
public abstract class PDFObject implements PDFWritable {

    /** logger for all PDFObjects (and descendants) */
    protected static Log log = LogFactory.getLog(PDFObject.class.getName());

    /**
     * the object's number
     */
    private int objnum;

    /**
     * the object's generation (0 in new documents)
     */
    private int generation = 0;

    /**
     * the parent PDFDocument
     */
    private PDFDocument document;

    /** the parent PDFObject (may be null and may not always be set, needed for encryption) */
    private PDFObject parent;

    /**
     * Returns the object's number.
     * @return the PDF Object number
     */
    public int getObjectNumber() {
        if (this.objnum == 0) {
            throw new IllegalStateException("Object has no number assigned: " + this.toString());
        }
        return this.objnum;
    }

    /**
     * Default constructor.
     */
    public PDFObject() {
        //nop
    }

    /**
     * Constructor for direct objects.
     * @param parent the containing PDFObject instance
     */
    public PDFObject(PDFObject parent) {
        setParent(parent);
    }

    /**
     * Indicates whether this PDFObject has already been assigned an
     * object number.
     * @return True if it has an object number
     */
    public boolean hasObjectNumber() {
        return this.objnum > 0;
    }

    /**
     * Sets the object number
     * @param objnum the object number
     */
    public void setObjectNumber(int objnum) {
        this.objnum = objnum;
        PDFDocument doc = getDocument();
        setParent(null);
        setDocument(doc); //Restore reference to PDFDocument after setting parent to null
        if (log.isTraceEnabled()) {
            log.trace("Assigning " + this + " object number " + objnum);
        }
    }

    /**
     * Returns the object's generation.
     * @return the PDF Object generation
     */
    public int getGeneration() {
        return this.generation;
    }

    /**
     * Returns the parent PDFDocument if assigned.
     * @return the parent PDFDocument (May be null if the parent PDFDocument
     * has not been assigned)
     */
    public final PDFDocument getDocument() {
        if (this.document != null) {
            return this.document;
        } else if (getParent() != null) {
            return getParent().getDocument();
        } else {
            return null;
        }
    }

    /**
     * Returns the parent PDFDocument, but unlike <code>getDocument()</code>
     * it throws an informative Exception if the parent document is unavailable
     * instead of having a NullPointerException somewhere without a message.
     * @return the parent PDFDocument
     */
    public final PDFDocument getDocumentSafely() {
        final PDFDocument doc = getDocument();
        if (doc == null) {
            throw new IllegalStateException("Parent PDFDocument is unavailable on "
                    + getClass().getName());
        }
        return doc;
    }

    /**
     * Sets the parent PDFDocument.
     * @param doc the PDFDocument.
     */
    public void setDocument(PDFDocument doc) {
        this.document = doc;
    }

    /**
     * Returns this objects's parent. The parent is null if it is a "direct object".
     * @return the parent or null if there's no parent (or it hasn't been set)
     */
    public PDFObject getParent() {
        return this.parent;
    }

    /**
     * Sets the direct parent object.
     * @param parent the direct parent
     */
    public void setParent(PDFObject parent) {
        this.parent = parent;
    }

    /**
     * Returns the PDF representation of the Object ID.
     * @return the Object ID
     */
    public String getObjectID() {
        return getObjectNumber() + " " + getGeneration() + " obj\n";
    }

    /**
     * Returns the PDF representation of a reference to this object.
     * @return the reference string
     */
    public String referencePDF() {
        if (!hasObjectNumber()) {
            throw new IllegalArgumentException(
                    "Cannot reference this object. It doesn't have an object number");
        }
        String ref = getObjectNumber() + " " + getGeneration() + " R";
        return ref;
    }

    /**
     * Creates and returns a reference to this object.
     * @return the object reference
     */
    public PDFReference makeReference() {
        return new PDFReference(this);
    }

    /**
     * Write the PDF represention of this object
     *
     * @param stream the stream to write the PDF to
     * @throws IOException if there is an error writing to the stream
     * @return the number of bytes written
     */
    protected int output(OutputStream stream) throws IOException {
        byte[] pdf = this.toPDF();
        stream.write(pdf);
        return pdf.length;
    }

    /** {@inheritDoc} */
    public void outputInline(OutputStream out, Writer writer) throws IOException {
        if (hasObjectNumber()) {
            writer.write(referencePDF());
        } else {
            writer.flush();
            output(out);
        }
    }

    /**
     * Encodes the object as a byte array for output to a PDF file.
     *
     * @return PDF string
     */
    protected byte[] toPDF() {
        return encode(toPDFString());
    }


    /**
     * This method returns a String representation of the PDF object. The result
     * is normally converted/encoded to a byte array by toPDF(). Only use
     * this method to implement the serialization if the object can be fully
     * represented as text. If the PDF representation of the object contains
     * binary content use toPDF() or output(OutputStream) instead. This applies
     * to any object potentially containing a string object because string object
     * are encrypted and therefore need to be binary.
     * @return String the String representation
     */
    protected String toPDFString() {
        throw new UnsupportedOperationException("Not implemented. "
                    + "Use output(OutputStream) instead.");
    }

    /**
     * Converts text to a byte array for writing to a PDF file.
     * @param text text to convert/encode
     * @return byte[] the resulting byte array
     */
    public static final byte[] encode(String text) {
        return PDFDocument.encode(text);
    }

    /**
     * Encodes a Text String (3.8.1 in PDF 1.4 specs)
     * @param text the text to encode
     * @return byte[] the encoded text
     */
    protected byte[] encodeText(String text) {
        if (getDocumentSafely().isEncryptionActive()) {
            final byte[] buf = PDFText.toUTF16(text);
            return PDFText.escapeByteArray(
                getDocument().getEncryption().encrypt(buf, this));
        } else {
            return encode(PDFText.escapeText(text, false));
        }
    }

    /**
     * Encodes a String (3.2.3 in PDF 1.4 specs)
     * @param string the string to encode
     * @return byte[] the encoded string
     */
    protected byte[] encodeString(String string) {
        return encodeText(string);
    }

    /**
     * Encodes binary data as hexadecimal string object.
     * @param data the binary data
     * @param out the OutputStream to write the encoded object to
     * @throws IOException if an I/O error occurs
     */
    protected void encodeBinaryToHexString(byte[] data, OutputStream out) throws IOException {
        out.write('<');
        if (getDocumentSafely().isEncryptionActive()) {
            data = getDocument().getEncryption().encrypt(data, this);
        }
        String hex = PDFText.toHex(data, false);
        byte[] encoded = hex.getBytes("US-ASCII");
        out.write(encoded);
        out.write('>');
    }

    /**
     * Formats an object for serialization to PDF.
     * @param obj the object
     * @param out the OutputStream to write to
     * @param writer a Writer for text content (will always be a wrapper around the above
     *                  OutputStream. Make sure <code>flush</code> is called when mixing calls)
     * @throws IOException If an I/O error occurs
     */
    protected void formatObject(Object obj, OutputStream out, Writer writer) throws IOException {
        if (obj == null) {
            writer.write("null");
        } else if (obj instanceof PDFWritable) {
            ((PDFWritable)obj).outputInline(out, writer);
        } else if (obj instanceof Number) {
            if (obj instanceof Double || obj instanceof Float) {
                writer.write(PDFNumber.doubleOut(((Number)obj).doubleValue()));
            } else {
                writer.write(obj.toString());
            }
        } else if (obj instanceof Boolean) {
            writer.write(obj.toString());
        } else if (obj instanceof byte[]) {
            writer.flush();
            encodeBinaryToHexString((byte[])obj, out);
        } else {
            writer.flush();
            out.write(encodeText(obj.toString()));
        }
    }

    /** Formatting pattern for PDF date */
    protected static final SimpleDateFormat DATE_FORMAT;

    static {
        DATE_FORMAT = new SimpleDateFormat("'D:'yyyyMMddHHmmss", Locale.ENGLISH);
        DATE_FORMAT.setTimeZone(TimeZone.getTimeZone("GMT"));
    }

    /**
     * Formats a date/time according to the PDF specification
     * (D:YYYYMMDDHHmmSSOHH'mm').
     * @param time date/time value to format
     * @param tz the time zone
     * @return the requested String representation
     */
    protected String formatDateTime(Date time, TimeZone tz) {
        Calendar cal = Calendar.getInstance(tz, Locale.ENGLISH);
        cal.setTime(time);

        int offset = cal.get(Calendar.ZONE_OFFSET);
        offset += cal.get(Calendar.DST_OFFSET);

        //DateFormat is operating on GMT so adjust for time zone offset
        Date dt1 = new Date(time.getTime() + offset);
        StringBuffer sb = new StringBuffer();
        sb.append(DATE_FORMAT.format(dt1));

        offset /= (1000 * 60); //Convert to minutes

        if (offset == 0) {
            sb.append('Z');
        } else {
            if (offset > 0) {
                sb.append('+');
            } else {
                sb.append('-');
            }
            int offsetHour = Math.abs(offset / 60);
            int offsetMinutes = Math.abs(offset % 60);
            if (offsetHour < 10) {
                sb.append('0');
            }
            sb.append(Integer.toString(offsetHour));
            sb.append('\'');
            if (offsetMinutes < 10) {
                sb.append('0');
            }
            sb.append(Integer.toString(offsetMinutes));
            sb.append('\'');
        }
        return sb.toString();
    }

    /**
     * Formats a date/time according to the PDF specification.
     * (D:YYYYMMDDHHmmSSOHH'mm').
     * @param time date/time value to format
     * @return the requested String representation
     */
    protected String formatDateTime(Date time) {
        return formatDateTime(time, TimeZone.getDefault());
    }

}

File: dotnet/src/Fop.Pdf/PDFObject.cs
using System.Globalization;
using System.Text;

namespace Fop.Pdf;

/// <summary>
/// Generic PDF object.
/// </summary>
/// <remarks>
/// A PDF document is essentially a collection of these objects. A PDF object has a number and a
/// generation (the generation is always 0 in new documents).
/// </remarks>
public abstract class PDFObject : IPDFWritable, ICompressedObject
{
    private bool _hasObjNum;
    private PDFObjectNumber _objNum = new();

    // The parent PDFDocument (typed as the minimal IPdfDocument abstraction).
    // TODO: full PDFDocument port.
    private IPdfDocument? _document;

    // The direct parent PDFObject (may be null; needed for encryption).
    private PDFObject? _parent;

    /// <summary>Default constructor.</summary>
    protected PDFObject()
    {
    }

    /// <summary>Constructor for direct objects.</summary>
    /// <param name="parent">The containing <see cref="PDFObject"/> instance.</param>
    protected PDFObject(PDFObject? parent) => Parent = parent;

    /// <summary>Returns the object's number.</summary>
    /// <exception cref="InvalidOperationException">If the object has no number assigned.</exception>
    public PDFObjectNumber ObjectNumber =>
        _hasObjNum
            ? _objNum
            : throw new InvalidOperationException("Object has no number assigned: " + ToString());

    /// <summary>Indicates whether this object has already been assigned an object number.</summary>
    public bool HasObjectNumber => _hasObjNum;

    /// <summary>Sets the object number, lazily allocated from the given document.</summary>
    public void SetObjectNumber(IPdfDocument document)
    {
        _objNum.SetDocument(document);
        _hasObjNum = true;
        IPdfDocument? doc = Document;
        Parent = null;
        Document = doc; // Restore reference to the document after setting parent to null.
    }

    /// <summary>Sets the object number from an existing <see cref="PDFObjectNumber"/>.</summary>
    public void SetObjectNumber(PDFObjectNumber objectNumber)
    {
        _objNum = objectNumber;
        _hasObjNum = true;
    }

    /// <summary>Sets the object number to a fixed value.</summary>
    public void SetObjectNumber(int objectNumber)
    {
        _objNum = new PDFObjectNumber(objectNumber);
        _hasObjNum = true;
    }

    /// <summary>Returns this object's generation. Always 0 in new documents.</summary>
    public int Generation => 0;

    /// <summary>
    /// The parent document, resolved through the parent chain if not set directly. May be
    /// <see langword="null"/> if no document has been assigned.
    /// </summary>
    public IPdfDocument? Document
    {
        get => _document ?? _parent?.Document;
        set => _document = value;
    }

    /// <summary>
    /// Returns the parent document, throwing an informative exception if it is unavailable
    /// (unlike <see cref="Document"/>, which returns <see langword="null"/>).
    /// </summary>
    public IPdfDocument DocumentSafely =>
        Document ?? throw new InvalidOperationException(
            "Parent PDFDocument is unavailable on " + GetType().FullName);

    /// <summary>
    /// This object's direct parent. Null for a "direct object" (or if it has not been set).
    /// </summary>
    public PDFObject? Parent
    {
        get => _parent;
        set => _parent = value;
    }

    /// <summary>Returns the PDF representation of the object ID, e.g. <c>"10 0 obj\n"</c>.</summary>
    public string ObjectID =>
        ObjectNumber + " " + Generation.ToString(CultureInfo.InvariantCulture) + " obj\n";

    /// <summary>Returns the PDF representation of a reference to this object.</summary>
    /// <exception cref="ArgumentException">If the object has no object number.</exception>
    public string ReferencePDF()
    {
        if (!HasObjectNumber)
        {
            throw new ArgumentException(
                "Cannot reference this object. It doesn't have an object number");
        }
        return MakeReference().ToString();
    }

    /// <summary>Creates and returns a reference to this object.</summary>
    public PDFReference MakeReference() => new(this);

    /// <summary>Writes the PDF representation of this object.</summary>
    /// <param name="output">The stream to write the PDF to.</param>
    /// <returns>The number of bytes written.</returns>
    public virtual int Output(Stream output)
    {
        byte[] pdf = ToPDF();
        output.Write(pdf, 0, pdf.Length);
        return pdf.Length;
    }

    /// <inheritdoc/>
    public virtual void OutputInline(Stream? output, StringBuilder textBuffer)
    {
        if (HasObjectNumber)
        {
            textBuffer.Append(ReferencePDF());
        }
        else
        {
            FlushTextBuffer(textBuffer, output!);
            Output(output!);
        }
    }

    /// <summary>Encodes the object as a byte array for output to a PDF file.</summary>
    protected virtual byte[] ToPDF() => Encode(ToPDFString());

    /// <summary>
    /// Returns a string representation of the PDF object. Only use this to implement serialization
    /// when the object can be fully represented as text; binary content must use
    /// <see cref="Output(Stream)"/> instead.
    /// </summary>
    protected virtual string ToPDFString() =>
        throw new NotSupportedException("Not implemented. Use Output(Stream) instead.");

    /// <summary>Converts text to a byte array for writing to a PDF file (ISO-8859-1).</summary>
    public static byte[] Encode(string text) => PdfEncoding.Encode(text);

    /// <summary>
    /// Flushes a text buffer to a stream with the right encoding and resets the buffer, so that
    /// binary content can follow. Exposed for <see cref="IPDFWritable"/> implementors.
    /// </summary>
    public static void FlushTextBuffer(StringBuilder textBuffer, Stream output) =>
        PdfEncoding.FlushTextBuffer(textBuffer, output);

    /// <summary>Encodes a Text String (3.8.1 in PDF 1.4 specs).</summary>
    protected byte[] EncodeText(string text)
    {
        // TODO: full PDFDocument port (encryption path). When encryption is active, FOP encrypts
        // the UTF-16 bytes; that collaborator is out of scope for this slice.
        if (DocumentSafely.IsEncryptionActive)
        {
            throw new NotSupportedException(
                "Encryption is not supported by this slice of the PDF object model.");
        }
        return Encode(PDFText.EscapeText(text, false));
    }

    /// <summary>Encodes a String (3.2.3 in PDF 1.4 specs).</summary>
    protected byte[] EncodeString(string str) => EncodeText(str);

    /// <summary>Encodes binary data as a hexadecimal string object.</summary>
    protected void EncodeBinaryToHexString(byte[] data, Stream output)
    {
        output.WriteByte((byte)'<');
        // TODO: full PDFDocument port (encryption path).
        if (DocumentSafely.IsEncryptionActive)
        {
            throw new NotSupportedException(
                "Encryption is not supported by this slice of the PDF object model.");
        }
        string hex = PDFText.ToHex(data, false);
        byte[] encoded = Encoding.ASCII.GetBytes(hex);
        output.Write(encoded, 0, encoded.Length);
        output.WriteByte((byte)'>');
    }

    /// <summary>
    /// Formats an object for serialization to PDF.
    /// </summary>
    /// <remarks>
    /// IMPORTANT: To write binary output, the text buffer is flushed first (see
    /// <see cref="FlushTextBuffer"/>) before writing any content to <paramref name="output"/>.
    /// </remarks>
    protected void FormatObject(object? obj, Stream? output, StringBuilder textBuffer)
    {
        switch (obj)
        {
            case null:
                textBuffer.Append("null");
                break;
            case IPDFWritable writable:
                writable.OutputInline(output, textBuffer);
                break;
            case double d:
                textBuffer.Append(PDFNumber.DoubleOut(d));
                break;
            case float f:
                textBuffer.Append(PDFNumber.DoubleOut(f));
                break;
            // All other numeric types serialize via their invariant string form, matching Java's
            // Number.toString() for the non-Double/Float case (e.g. Integer, Long).
            case sbyte or byte or short or ushort or int or uint or long or ulong:
                textBuffer.Append(Convert.ToInt64(obj, CultureInfo.InvariantCulture)
                    .ToString(CultureInfo.InvariantCulture));
                break;
            case bool b:
                // Java Boolean.toString() emits lowercase "true"/"false".
                textBuffer.Append(b ? "true" : "false");
                break;
            case byte[] bytes:
                FlushTextBuffer(textBuffer, output!);
                EncodeBinaryToHexString(bytes, output!);
                break;
            default:
                FlushTextBuffer(textBuffer, output!);
                byte[] encoded = EncodeText(obj.ToString() ?? string.Empty);
                output!.Write(encoded, 0, encoded.Length);
                break;
        }
    }

    /// <summary>
    /// Checks whether the other object has the same content. The contract is weaker than
    /// <see cref="object.Equals(object)"/>: it need not check the object ID.
    /// </summary>
    protected virtual bool ContentEquals(PDFObject? o) => Equals(o);

    /// <summary>Collects child objects (no-op by default).</summary>
    public virtual void GetChildren(ISet<PDFObject> children)
    {
    }

    /// <summary>Indicates whether this object may appear in an object stream.</summary>
    public virtual bool SupportsObjectStream => true;
}
